specification/dartLangSpec.tex

```diff
@@ -11829,7 +11829,40 @@ \subsection{Function Expressions}
 
 \rationale{%
 This definition guarantees that for any type $T$,
-\code{$T <:$ FutureOr<$\flatten{T}$>}.%
+\code{$T <:$ FutureOr<$\flatten{T}$>}. The proof is by induction on `T`:
+
+\begin{itemize}
+\item If $T$ is \code{$X$\,\&\,$S$} then
+  \begin{itemize}
+  \item if $S$ derives a future type $U$,
+    then \code{$T <: S$} and \code{$S <: U$}, so \code{$T <: U$}.
```

Comment:
Do we need to show that `S <: U` inductively on the definition of "derives a future type"?

Or generally that if `T` derives a future type `U` then `T <: U (if that is true... I think it is.)

Reply:
The spec notes this property after the definition of "derives", but the casework is straightforward:
* Case 1: `T` is a type introduced by a class, mixin, or enum declaration, and `T` or some superinterface of `T` is `Future<S>` for some `S`, and `U = Future<S>`. Then `T <: U` is trivial.
* Case 2: `T` is `FutureOr<S>` for some `S` and `U = FutureOr<S>`. `T <: U` is again trivial.
* Case 3: `T` is `S?` for some `S`, `S` derives the future type `F`, and `U = F?`. Then to show `T <: U`, we must show `S? <: F?`. We have `S <: F` by induction, so `S? <: F?` follows by the usual rules for nullable types.
* Case 4: `T` is a type variable with bound `B`, `B` derives the future type `F`, and `U = F`. Then `B <: F` by induction, so `T <: B <: F = U`.

Reply:
@fishythefish agreed! While you were typing this up, I was modifying my PR, adding a similar argument to the spec.

I _think_ our arguments are equivalent. Let me know if you agree.

Reply:
Yup, I think we're in agreement.